cupy/indexing/generate.py

```diff
@@ -1,13 +1,163 @@
 # flake8: NOQA
 # "flake8: NOQA" to suppress warning "H104  File contains nothing but comments"
 
-
-# class c_(object):
-# class r_(object):
 # class s_(object):
 
 import numpy
+import six
+
 import cupy
+from cupy import core
+from cupy.creation import from_data
+from cupy.manipulation import join
+
+
+class AxisConcatenator(object):
+    """Translates slice objects to concatenation along an axis.
+
+    For detailed documentation on usage, see :func:`cupy.r_`.
+    This implementation is partially borrowed from NumPy's one.
+
+    """
+
+    def _output_obj(self, obj, ndim, ndmin, trans1d):
+        k2 = ndmin - ndim
+        if trans1d < 0:
+            trans1d += k2 + 1
+        defaxes = list(six.moves.range(ndmin))
+        k1 = trans1d
+        axes = defaxes[:k1] + defaxes[k2:] + \
+            defaxes[k1:k2]
+        return obj.transpose(axes)
+
+    def __init__(self, axis=0, matrix=False, ndmin=1, trans1d=-1):
```

Comment:
Are these default arguments really required? They are overwritten by its subclass. Users may use this class directly?